File: 윤호선/25주차/다단계_칫솔_판매.cpp
/*
    문제를 잘 읽자.
*/
#include <string>
#include <vector>
#include <map>
using namespace std;

map<string, int> member_num;
int cal(int num, int s, vector<string>& referral, vector<int>& answer){
    int p=s*0.1;
    answer[num]+=s;
    answer[num]-=p;
    if(referral[num]=="-")return 0;
    if(p==0)return 0;
    cal(member_num[referral[num]], p, referral, answer);
    return 0;
}
vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    vector<int> answer(enroll.size());
    
    for(int i=0;i<enroll.size();i++){
        member_num.insert({enroll[i],i});
    }
    for(int i=0;i<seller.size();i++){    
        cal(member_num[seller[i]],amount[i]*100, referral, answer);
    }     
    return answer;
}

File: 노영주/25주차/다단계칫솔판매.js
function solution(enroll, referral, seller, amount) {
  const group = {};
  const profit = {};

  enroll.forEach((name) => {
    group[name] = [];
    profit[name] = 0;
  });

  referral.forEach((name, idx) => {
    group[enroll[idx]].push(name);
  });

  const distributionProfit = (name, price) => {
    if (price === 0) {
      return;
    }
    if (name === "-") {
      return;
    }

    const distribution = Math.floor(price * 0.1);
    profit[name] += price - distribution;
    distributionProfit(group[name], distribution);
  };

  seller.forEach((name, idx) => {
    const price = amount[idx] * 100;
    distributionProfit(name, price);
  });

  return Array.from(enroll.map((name) => profit[name]));
}
